src/NUnitFramework/framework/Internal/Reflect.cs

```diff
@@ -274,6 +281,83 @@ internal static bool CanImplicitlyConvertTo(this Type? from, Type to)
             }
         }
 
+        /// <summary>
+        /// Invokes <see cref="IConstraint.ApplyTo{TActual}(TActual)"/> using the given <paramref name="constraint"/> as the instance
+        /// with <paramref name="actualType"/> as its generic type argument.
+        /// </summary>
+        /// <param name="constraint">The <see cref="IConstraint"/>.</param>
+        /// <param name="actualType">The generic type argument to pass to <see cref="IConstraint.ApplyTo{TActual}(TActual)"/>.</param>
+        /// <param name="actual">The actual value.</param>
+        /// <returns>A <see cref="ConstraintResult"/>.</returns>
+        public static ConstraintResult InvokeApplyTo(IConstraint constraint, Type? actualType, object? actual)
+        {
+            Guard.ArgumentNotNull(constraint, nameof(constraint));
+
+            actualType ??= typeof(object);
+
+            Guard.ArgumentValid(actual is null
+                ? IsAssignableFromNull(actualType)
+                : actualType.IsAssignableFrom(actual.GetType()),
+                "The actual value must be assignable to the provided actual type.",
+                nameof(actual));
+
+            static ConstraintResult DelegateTemplate<T>(IConstraint constraint, object? actual)
+                => constraint.ApplyTo((T?)actual);
+
+            static Func<IConstraint, object?, ConstraintResult> BuildDelegate(Type type)
+                => ((Delegate)DelegateTemplate<object>).Method
+                .GetGenericMethodDefinition()
+                .MakeGenericMethod(type)
+                .CreateDelegate<Func<IConstraint, object?, ConstraintResult>>();
+
+            return InvokeApplyToLookup.GetOrAdd(actualType, BuildDelegate).Invoke(constraint, actual);
+        }
+
+        /// <summary>
+        /// Invokes <see cref="ICollectionConstraint.ApplyToCollection{TActual, TItem}(TActual, IEnumerable{TItem})"/>
+        /// using the given <paramref name="constraint"/> as the instance with <paramref name="actualType"/> and
+        /// <paramref name="itemType"/> as its generic type arguments.
+        /// </summary>
+        /// <param name="constraint">The <see cref="IConstraint"/>.</param>
+        /// <param name="actualType">The actual type argument to pass to <see cref="ICollectionConstraint.ApplyToCollection{TActual, TItem}(TActual, IEnumerable{TItem})"/>.</param>
+        /// <param name="itemType">The item type argument to pass to <see cref="ICollectionConstraint.ApplyToCollection{TActual, TItem}(TActual, IEnumerable{TItem})"/>.</param>
+        /// <param name="actual">The actual value.</param>
+        /// <param name="collection">The collection.</param>
+        /// <returns>A <see cref="ConstraintResult"/>.</returns>
+        public static ConstraintResult InvokeApplyToCollection(ICollectionConstraint constraint, Type? actualType, Type? itemType, object? actual, IEnumerable collection)
```

Comment:
Again `actualType` cannot be `null`, nor can `actual` (which is the same as _collection_).
All constraints that call this have a `Guard` on `IEnumerable` with _allowNull_ set to `false.
Neither are _actual_ nor _itemType_.